Support backwards compatible reads for unnanotated repeated primitive fields in Parquet

mxmarkovics opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->
<!-- Provide an overview of the PR for maintainers and reviewers. -->
## Description
Per the [Parquet Spec](https://github.com/apache/parquet-format/blob/master/LogicalTypes.md): `A repeated field that is neither contained by a LIST- or MAP-annotated group nor annotated by LIST or MAP should be interpreted as a required list of required elements where the element type is the type of the field`. So the following schemas should be treated equivalently.

```
# Old schema format for required list of required ints
message TestProtobuf.RepeatedIntMessage {
  repeated int32 repeatedInt;
}
# New schema format for required list of required ints
message TestProtobuf.RepeatedIntMessage {
  required group repeatedInt (LIST) {
    repeated group list {
      required int32 element;
    }
  }
}
```
However, Trino currently does not support reading the old format and will throw the following error if attempting to read the field as a `array<int>`:
```
Query 20240229_210455_00000_hegy2 failed: Error opening Hive split s3://...: class org.apache.parquet.io.PrimitiveColumnIO cannot be cast to class org.apache.parquet.io.GroupColumnIO (org.apache.parquet.io.PrimitiveColumnIO and org.apache.parquet.io.GroupColumnIO are in unnamed module of loader io.trino.server.PluginClassLoader @56b37c54)
```
If instead the field is defined in the table as simply the primitive type itself e.g. `int` then another unhelpful error will be thrown:
```
GENERIC_INTERNAL_ERROR: Loaded block positions count (3) doesn’t match lazy block positions count (1)
```
This issue is also present in Presto, however the second case will actually return incorrect results instead of throwing the second error.
Spark is able to correctly handle this however, e.g.:
```
inputDF = spark.read.parquet(“s3://...“).select(“*”);
inputDF.show()
...
+-----------+
|repeatedInt|
+-----------+
|  [1, 2, 3]     |
+-----------+
```
Have added two old parquet file that was written with the old format showing the reproduction and used for unit testing. Tested on AWS Athena.
There are several issues here:
1. In `ParquetTypeUtils` the table declared type `array<int>` assumes that the Parquet `ColumnIO` will be a `GroupColumIO`, however in this case the `ColumnIO` is a `PrimitiveColumnIO` which is repeated.
2. The above issue aside, Trino should interpret the repeated non-group field Parquet schema as a required list of required elements per the spec. However, when wrapping the `PrimitiveField` with a `GroupField` as it would be in the case of the new schema format causes the `GroupField` to have the same repetition and definition levels of the `PrimitiveField` child, where it should have 1 less for each since it is one level above. This change is also consistent with the change made in Spark to handle this case as well: [https://github.com/apache/spark/commit/d246010b8b413868c2cad28b6d46e14075d841a4](https://github.com/apache/spark/commit/d246010b8b413868c2cad28b6d46e14075d841a4).
3. When the declared table type is incompatible with the repeated field in the schema an unhelpful error message is thrown that does not indicate there is a schema mismatch, and in the case of Presto will give incorrect results. This should only be the case when the repeated primitive field as at the top level, i.e. parent is a MessageColumnIO, to still be backwards compatible with old Avro format as well.

Note: There is a similar code path in the Iceberg connector in `IcebergParquetColumnIOConverter`, however I was unable to reproduce this issue using Iceberg as it currently does not allow this case, so we don't need to handle the Iceberg case in Trino. [https://github.com/apache/iceberg/blob/main/parquet/src/main/java/org/apache/iceberg/parquet/MessageTypeToType.java#L74-L77](https://github.com/apache/iceberg/blob/main/parquet/src/main/java/org/apache/iceberg/parquet/MessageTypeToType.java#L74-L77).


<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Hive
* Fix reading array types from parquet files produced by some legacy writes  ({issue}`20943`)
```

pettyjamesm commented on lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java:
```diff
@@ -354,14 +355,37 @@ public static Optional<Field> constructField(Type type, ColumnIO columnIO)
+            //              }
+            //         }
+            //      2. repeated binary colors (STRING);
+            if (required && columnIO instanceof PrimitiveColumnIO && repetitionLevel > 0 && definitionLevel > 0) {
```
Also requires `columnIO.getType().getRepetition() == REPEATED`

mxmarkovics replied:
Yeah, mistakenly replaced the repeated check you described with the `required` check, but should be the opposite.

mxmarkovics commented:
Fixed failing unit tests by adding an additional check in `ColumnReaderFactory`. The tests were failing because the batched reader used the `FlatColumnReader` instead of the `NestedColumnReader` via `ColumnReaderFactory` for repeated top level primitive fields, however the primitive field in the normal/new schema format for arrays it would use `NestedColumnReader` . In other words, the `FlatColumnReader` is only used to read top level primitive fields, however since top level repeated primitives should be treated the same as nested primitives inside of a list group, they should be considered as nested instead of flat columns. The additional check in `isFlatColumns` should work because the repetition and definition levels for top level repeated primitives are (1, 1) and the only case where a top level primitive can have a max repetition level of 1 is if itself is repeated. Nested repeated primitives of other forms are handled already in `NestedColumnReader` via handling for the old Avro format and the normal/current way of representing lists in the schema.

Other changes:
* Improved error message for schema mismatch
* Added unit test coverage for error message
* Added unit test coverage for change to `isFlatColumn`

pettyjamesm commented on lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java:
```diff
@@ -354,14 +355,38 @@ public static Optional<Field> constructField(Type type, ColumnIO columnIO)
         PrimitiveColumnIO primitiveColumnIO = (PrimitiveColumnIO) columnIO;
+        if (primitiveColumnIO.getType().getRepetition() == REPEATED && primitiveColumnIO.getParent() instanceof MessageColumnIO) {
+            ColumnDescriptor columnDescriptor = primitiveColumnIO.getColumnDescriptor();
+            throw new TrinoException(TYPE_MISMATCH, format("Repeated field %s's type %s in a parquet file is incompatible with type %s defined in table schema",
```
I would match what `ColumnReaderFactory.unsupportedException` throws

pettyjamesm commented on lib/trino-parquet/src/main/java/io/trino/parquet/ParquetTypeUtils.java:
```diff
@@ -354,14 +355,38 @@ public static Optional<Field> constructField(Type type, ColumnIO columnIO)
+                PrimitiveField primitiveFieldElement = new PrimitiveField(((ArrayType) type).getElementType(), true, primitiveColumnIO.getColumnDescriptor(), primitiveColumnIO.getId());
+                return Optional.of(new GroupField(type, repetitionLevel - 1, definitionLevel - 1, true, ImmutableList.of(Optional.of(primitiveFieldElement))));
+            }
+            else {
```
No need to wrap in an `else` block here since the `if` statement returns at the end.

pettyjamesm commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java:
```diff
@@ -1988,6 +1995,85 @@ public void testMapMaxReadBytes()
+                            .row(Arrays.asList("good", "bye"))
+                            .row(Arrays.asList("one", "two", "three"))
+                            .build());
+            queryRunner.execute("DROP TABLE table_with_repeated_primitive_string");
```
Move to a `finally` block unless this all goes away as a result of using a a page source directly for testing.

pettyjamesm commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java:
```diff
@@ -1988,6 +1995,85 @@ public void testMapMaxReadBytes()
+                    .result().matches(resultBuilder(queryRunner.getDefaultSession(), new ArrayType(INTEGER))
+                            .row(Arrays.asList(1, 2, 3))
+                            .build());
+            queryRunner.execute("DROP TABLE table_with_repeated_primitive_int");
```
Move to a `finally` block unless this all goes away as a result of using a a page source directly for testing.

pettyjamesm commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java:
```diff
@@ -1988,6 +1995,85 @@ public void testMapMaxReadBytes()
+        finally {
+            queryRunner.execute("DROP TABLE IF EXISTS table_with_repeated_primitive_int_bad_table_schema");
+        }
+        assert errorThrown;
```
use `assertThatThrownBy(() -> ...)` instead, similar to what you see in `AbstractTestParquetReader#testParquetShortDecimalWriteToTrinoBigintBlockWithNonZeroScale` that includes assertions on the message and exception type.

pettyjamesm reviewed (commented):
The approach looks good to me, @raunaqmorarka PTAL

raunaqmorarka commented on plugin/trino-hive/src/test/java/io/trino/plugin/hive/parquet/AbstractTestParquetReader.java:
```diff
@@ -1988,6 +1990,35 @@ public void testMapMaxReadBytes()
         tester.testMaxReadBytes(getStandardMapObjectInspector(javaStringObjectInspector, javaLongObjectInspector), values, values, mapType(VARCHAR, BIGINT), maxReadBlockSize);
     }
 
+    @Test
+    public void testBackwardsCompatibleRepeatedStringField()
+            throws Exception
+    {
+        File parquetFile = new File(Resources.getResource("parquet_repeated_primitives/string/old-repeated-string.parquet").toURI());
+        List<List<String>> expectedValues = ImmutableList.of(Arrays.asList("hello", "world"), Arrays.asList("good", "bye"), Arrays.asList("one", "two", "three"));
+        tester.testReadPageOnly(parquetFile, ImmutableList.of("myString"), new ArrayType(VARCHAR), expectedValues);
+    }
+
+    @Test
+    public void testBackwardsCompatibleRepeatedIntegerField()
+            throws Exception
+    {
+        File parquetFile = new File(Resources.getResource("parquet_repeated_primitives/int/old-repeated-int.parquet").toURI());
+        List<List<Integer>> expectedValues = ImmutableList.of(Arrays.asList(1, 2, 3));
+        tester.testReadPageOnly(parquetFile, ImmutableList.of("repeatedInt"), new ArrayType(INTEGER), expectedValues);
+    }
+
+    @Test
+    public void testBackwardsCompatibleRepeatedPrimitiveFieldDefinedAsPrimitive()
+    {
+        assertThatThrownBy(() -> {
+            File parquetFile = new File(Resources.getResource("parquet_repeated_primitives/int/old-repeated-int.parquet").toURI());
+            List<List<Integer>> expectedValues = ImmutableList.of(Arrays.asList(1, 2, 3));
+            tester.testReadPageOnly(parquetFile, ImmutableList.of("repeatedInt"), INTEGER, expectedValues);
+        }).hasMessage("Unsupported Trino column type (integer) for Parquet column ([repeatedint] repeated int32 repeatedint)")
+                .isInstanceOf(TrinoException.class);
+    }
```
These tests should go into trino-parquet module rather than trino-hive as we don't use hive/hadoop libs to write the data here. You can either add a new class or add these to `io.trino.parquet.reader.TestParquetReader`

mxmarkovics replied:
Sure, moved the tests from `AbstractTestParquetReader` and the helper method in `ParquetTester` in trino-hive to `TestParquetReader` in trino-parquet.

raunaqmorarka reviewed (commented):
lgtm % comment

pettyjamesm reviewed (approved):
Looks good to me, thanks @mxmarkovics !
